### ISSUE TITLE
Inconsistent state using persist feature

### ISSUE THREAD

**Issue by sirpi** [open-policy-agent/opa#5428]
## Short description 
tl;dr: The agent does not update its state at some point when the persist feature is switched on. Occurred with version 0.46.1, both in Kubernetes and Windows. 

More detailed descriprtion: 
We use multiple bundles (currently around 30) with discovery, and on top of that, there are dependencies between them: there is a utility bundle with policies containing helper methods only, and the policies in the other bundles import these helper policies.
The persist feature has been switched on for all bundles, otherwise the initial response time may be more than a minute after an agent restart (with persist it is only 5-10 seconds). Recently we realized that the agent started responding based on outdated policies. It turned out the the outdated state is exactly the one persisted and for some unknown reason the agent stopped updating the persisted state at some point.

During agent restart there are always errors ("Bundle activation failed: 11 errors occurred:\nresources.a.a.rego:8: rego_type_error: undefined function ...") until the utility bundle is loaded, but it is normal since most of the policies refer to this bundle. But after the utility bundle is successfully loaded the errors disappear and the remaining bundles are updated. But in the inconsistent case the agent continuously complained about undefined functions and never recovered. First we suspected that the issue is with the Kubernetes environment when the 3 agent instances update the same files on the persistent storage, but it also occurred locally with a Windows persist folder and with a single instance, so the issue is somewhere else.

There is no authentication or network issue, the agent can download the bundles based on the log.

The system has been updated recently to 46.1 and before that we haven't experienced this behavior (the persist feature is switched on for months now).

Deleting the persist directory and restarting the agent solves the issue.

Sorry, I know this description is quite vague, but I tried to reproduce it with only two bundles when one refers to the other by e.g. updating the content of both bundles, or 1. stopped the agent, 2. update the bundles again, 3. restart the agent - without any luck. Do you have any idea about this behavior?

## Steps To Reproduce
Unfortunately I did not find a consistent way to reproduce the issue yet.

## Expected behavior
The bundles should be persisted after each successful bundle download.

## Additional context
N/A

**Comment by srenatus** [comment]
Thanks for the report. I'm afraid it's really hard to dig into this with this little to go on, though.

Would it be possible to turn on debug logging for one of your nodes, and see if it happens with that enabled...?

**Comment by ashutosh-narkar** [comment]
Closing this. Feel free to re-open if you notice this again and can help reproduce. Thanks!
